mapper/pojo-base/src/main/java/org/hibernate/search/mapper/pojo/mapping/definition/annotation/processing/impl/AbstractFieldAnnotationProcessor.java

```diff
@@ -46,8 +46,15 @@ public final void process(PropertyMappingStep mappingContext, A annotation,
 				valueBinder,
 				context
 		);
+		Map<String, Object> params = context.toMap( annotation );
+		for ( Map.Entry<String, Object> entry : context.toMap( valueBinder.params() ).entrySet() ) {
+			Object overriddenValue = params.put( entry.getKey(), entry.getValue() );
+			if ( overriddenValue != null ) {
+				log.annotationAttributeOverriddenWithParameter( annotation.annotationType(), entry.getKey(), overriddenValue,
+						entry.getValue() );
+			}
+		}
```

Comment:
> They need to know about the number of dimensions, because that’s a mandatory parameter of IndexFieldTypeFactory#asVector…  but they don't have access to the number passed to @VectorField#dimension.

the idea here that we just give access to all the attributes from the annotation as parameters to the binder, and then binder implementers can do what they want with the info...

Reply:
I'm not sure this is a good idea, as it essentially breaks the clear separation we used to have between annotations and the rest of the code, which allowed it to work in contexts where Java annotations don't make sense (e.g. Infinispan schema definitions).

As for suggestions... IDK, I was thinking of simply making `dimensions()` optional in the annotation and in `IndexFieldTypeFactory`, moving it back to an optional parameter. We could handle it a bit like we handle `decimalScale`.
Then binders would be able to set the number dimensions, and mapping (through annotations or programmatic API) would be able to set it explicitly when the binder doesn't (or even to override the binder's choice, though I doubt that is useful...).
Admittedly that would not be a great API but at least it would be consistent with the existing APIs and behaviors. 

Ideally the number of dimensions should really be part of the type: we don't map to a `float[]`, but to a `@Dimensions(10) float[]`. I can imagine an API supporting something like that (e.g. `implements ValueBridge<ImageContent, @Dimensions(1024) float[]>`, `private @Dimensions(1024) float[] embeddings;`, ...). We could even handle it right now with "synthetic" subtypes of `float[]` represented by a `RawTypeModel` that has some additional metadata attached. But I suspect this would have side-effects and require extensive care to make sure edge cases work correctly and/or lead to clear error messages, including in places where we actually don't care about vectors.... So, probably an idea best kept on the shelve :x

`decimalScale()` is a bit the same thing as `dimensions()`, so we might as well handle dimensions the same way... I guess?

Reply:
Right ... if we don't add new API, then we already have dimensions as optional on the annotation as well as on `PropertyMappingStep#vectorField`, so that part is covered? I was thinking more about the imaginary case 😄 
 where someone would add dimension to the annotation and then also provide a custom binder, and would want to somehow access that dimension value from the annotation. And for those who do not rely on annotations, it would be as is (just using the programmatic API, I assume?)

Reply:
> where someone would add dimension to the annotation and then also provide a custom binder, and would want to somehow access that dimension value from the annotation

Unsupported, they should use a custom annotation and annotation processor that passes that info to the binder.

> And for those who do not rely on annotations, it would be as is (just using the programmatic API, I assume?)

Not sure what you mean by "as is".

Reply:
> Not sure what you mean by "as is".

I meant that this change was only for the annotation processor, so only relevant for those who are using the annotations, and others who don't use them would not notice any difference 🙈 
But then if you are saying that in this case a user should just create a custom annotation, then I'll just drop this commit 😃

Reply:
> I meant that this change was only for the annotation processor, so only relevant for those who are using the annotations, and others who don't use them would not notice any difference 🙈

Yes, people who use the programmatic API actually don't have this problem as they can pass information directly through the binder's constructor.

>  But then if you are saying that in this case a user should just create a custom annotation, then I'll just drop this commit 😃

:+1: